server/aesdsocket.c

```diff
@@ -168,353 +188,425 @@ void closed_connection(struct sockaddr_storage client_address)
-                int lock_ret = pthread_mutex_lock(thread_data->mutex);
-                if(lock_ret != 0){
-                    syslog(LOG_ERR, "Error occured during mutex lock = %s. Exiting...", strerror(errno));
-                    goto exit_label3;  
+            //Check for whitespace
+            if(recv_data[index-1] == '\n'){
+                //Acquire mutex lock
+                common_retval = pthread_mutex_lock(thread_data->mutex);;
+                if(common_retval != 0){
+                    syslog(LOG_ERR, "Error occured while acquiring mutex lock = %s. Exiting...\n", strerror(errno));
+                    goto exit_label3;
                 }
 
-                //Write to logfile         
-                if(lseek(thread_data->logfile_fd, 0, SEEK_END) == -1){
-                    syslog(LOG_ERR, "Error occured during lseek() to EOF = %s. Exiting...", strerror(errno));
-                    int unlock_ret = pthread_mutex_unlock(thread_data->mutex);
-                    if(unlock_ret != 0){
-                        syslog(LOG_ERR, "Error occured during mutex lock release = %s. Exiting...", strerror(errno));
-                    }
-                    goto exit_label3;  
+#ifndef USE_AESD_CHAR_DEVICE
+                if(lseek(file_fd, 0, SEEK_END) == -1){
+                    syslog(LOG_ERR, "Error occured trying to reach EOF = %s. Exiting...\n", strerror(errno));
+                    //Release mutex lock
+                    common_retval = pthread_mutex_unlock(thread_data->mutex);
+                    if(common_retval != 0)
+                        syslog(LOG_ERR, "Error occured during releasing mutex lock = %s. Exiting..\n", strerror(errno));
+
+                    goto exit_label3;
                 }
 
-                int written_bytes;
-                int len_to_write = index;
-                char *ptr_to_write = recv_data;
-                while(len_to_write != 0){
-                    written_bytes = write(thread_data->logfile_fd, ptr_to_write, len_to_write);
-                    if(written_bytes == -1){
-                        if(errno == EINTR){ //User gave Ctrl+C
-                            int unlock_ret = pthread_mutex_unlock(thread_data->mutex);
-                            if(unlock_ret != 0){
-                                syslog(LOG_ERR, "Error occured during mutex lock release in write() = %s. Exiting...", strerror(errno));
+#else           
+                file_fd = open(LOGFILE_PATH, O_CREAT | O_RDWR | O_TRUNC, 0777);
+                if(file_fd == -1){
+                    syslog(LOG_ERR, "Error occured creating and opening logfile = %s. Exiting...\n", strerror(errno));
+                    goto exit_label3;
+                }
+
+#endif
+            //Commond for IOCTL seek
+            if(strncmp(recv_data, IOCTL_COMMAND, COMMAND_LEN) == 0){
+                struct aesd_seekto seek_to;
+                sscanf(recv_data, "AESDCHAR_IOCSEEKTO:%d,%d", &seek_to.write_cmd, &seek_to.write_cmd_offset);
```

Comment:
@prvbharadwaj1731 you should add error handling here, see discussion in class about this.